x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/kerberos/KerberosRealm.java

```diff
@@ -0,0 +1,190 @@
+        }
+        this.kerberosTicketValidator = kerberosTicketValidator;
+        this.threadPool = threadPool;
+    }
+
+    @Override
+    public Map<String, List<String>> getAuthenticationFailureHeaders() {
+        return Collections.singletonMap(WWW_AUTHENTICATE, Collections.singletonList(NEGOTIATE_SCHEME_NAME));
+    }
+
+    @Override
+    public void expire(final String username) {
+        if (userPrincipalNameToUserCache != null) {
+            userPrincipalNameToUserCache.invalidate(username);
+        }
+    }
+
+    @Override
+    public void expireAll() {
+        if (userPrincipalNameToUserCache != null) {
+            userPrincipalNameToUserCache.invalidateAll();
+        }
+    }
+
+    @Override
+    public boolean supports(final AuthenticationToken token) {
+        return token instanceof KerberosAuthenticationToken;
+    }
+
+    @Override
+    public AuthenticationToken token(final ThreadContext context) {
+        return KerberosAuthenticationToken.extractToken(context.getHeader(AUTH_HEADER));
+    }
+
+    @Override
+    public void authenticate(final AuthenticationToken token, final ActionListener<AuthenticationResult> listener) {
+        if (token instanceof KerberosAuthenticationToken) {
+            final KerberosAuthenticationToken kerbAuthnToken = (KerberosAuthenticationToken) token;
+            final Path keytabPath =
+                    config.env().configFile().resolve(KerberosRealmSettings.HTTP_SERVICE_KEYTAB_PATH.get(config.settings()));
+            final boolean krbDebug = KerberosRealmSettings.SETTING_KRB_DEBUG_ENABLE.get(config.settings());
+
+            kerberosTicketValidator.validateTicket((byte[]) kerbAuthnToken.credentials(), keytabPath, krbDebug,
+                    ActionListener.wrap((userPrincipalNameOutToken) -> {
+                        if (userPrincipalNameOutToken.v1() != null) {
+                            buildUser(userPrincipalNameOutToken, listener);
+                        } else {
+                            /**
+                             * Ongoing context establishment, terminate with UNAUTHORIZED and outToken
+                             */
+                            String errorMessage = "failed to authenticate user, gss context negotiation not complete";
+                            listener.onResponse(AuthenticationResult.terminate(errorMessage,
+                                    unauthorized(errorMessage, null, userPrincipalNameOutToken.v2())));
+                        }
+                    }, (e) -> {
+                        handleException(e, listener);
+                    }));
+        } else {
+            listener.onResponse(AuthenticationResult.notHandled());
+        }
+    }
+
+    private void handleException(Exception e, final ActionListener<AuthenticationResult> listener) {
+        if (e instanceof LoginException) {
+            logger.error("failed to authenticate user, service login failure", e);
```

Comment:
I don't think we want to log an `error` everytime a login fails do we?
It seems quite verbose, especially since the `AuthenticationService` will log this as well.